external/fv3fit/fv3fit/sklearn/_min_max_novelty_detector.py

```diff
@@ -0,0 +1,131 @@
+from fv3fit._shared import stacking
+from .. import _shared
+from .._shared import (
+    match_prediction_to_input_coords,
+    pack,
+    pack_tfdataset,
+    Predictor,
+    register_training_function,
+    SAMPLE_DIM_NAME,
+    stack_non_vertical,
+)
+from .._shared.config import MinMaxNoveltyDetectorHyperparameters, PackerConfig
+
+import dataclasses
+import dacite
+import fsspec
+from fv3fit import tfdataset
+from fv3fit._shared.packer import PackingInfo, clip_sample, unpack
+from fv3fit.tfdataset import apply_to_mapping, ensure_nd
+from fv3fit.typing import Batch
+import io
+import joblib
+import logging
+import numpy as np
+from sklearn.preprocessing import MinMaxScaler
+import tensorflow as tf
+from typing import Optional, Iterable, Sequence
+from vcm import safe
+import xarray as xr
+import yaml
+
+@register_training_function("min_max_novelty_detector", MinMaxNoveltyDetectorHyperparameters)
+def train_min_max_novelty_detector(
+    hyperparameters: MinMaxNoveltyDetectorHyperparameters,
+    train_batches: tf.data.Dataset,
+    validation_batches: tf.data.Dataset
+):
+    train_batches = train_batches.map(
+        tfdataset.apply_to_mapping(tfdataset.float64_to_float32)
+    )
+    
+    model = MinMaxNoveltyDetector(hyperparameters)
+    model.fit(train_batches)
+    return model
+
+@_shared.io.register("minmax")
+class MinMaxNoveltyDetector(Predictor):
+
+    _PICKLE_NAME = "minmax.pkl"
+    _METADATA_NAME = "metadata.bin"
+
+    _NOVELTY_OUTPUT_VAR = "is_novelty"
+
+    dims = ["x", "y", "tile", "time"]
+
+    def __init__(
+        self,
+        hyperparameters: MinMaxNoveltyDetectorHyperparameters
+    ):
+        output_variables = [self._NOVELTY_OUTPUT_VAR]
+        super().__init__(hyperparameters.input_variables, output_variables)
+        self.packer_config = hyperparameters.packer_config
+
+    def fit(self, batches: tf.data.Dataset):
+        logger = logging.getLogger("MinMaxNoveltyDetector")
+        logger.info(f"Fitting min-max novelty detector.")
+
+        batches = batches.map(apply_to_mapping(ensure_nd(2)))
+        batches_clipped = batches.map(clip_sample(self.packer_config.clip))
+        x_dataset, self.input_features_ = pack_tfdataset(
+            batches_clipped, [str(item) for item in self.input_variables]
+        )
+        x_dataset = x_dataset.unbatch()
+
+        total_samples = sum(1 for _ in iter(x_dataset))
+        X: Batch = next(iter(x_dataset.batch(total_samples)))
+
+        self.scaler = MinMaxScaler()
+        self.scaler.fit(X)
+
+        logger.info(f"Min-max novelty detector done fitting.")
+
+    def predict(self, data: xr.Dataset) -> xr.Dataset:  
```

Comment:
Thank you, that's a great suggestion! The only ways in which I've currently deviated are (1) having the output of `.predict` be an `xr.Dataset` rather than the direct output of the sklearn class (so that we can specify the output for every novelty detector and that we can use the same `predict_novelty` method for each) and (2) have 1/0 outputs rather than Booleans, since it seems somewhat easier to deal with for aggregation. I'm flexible on either.